File: forklift/models/schema.py
#!/usr/bin/env python
# * coding: utf8 *
"""
schema.py
A module that holds the schema for the dat file
"""

from collections import OrderedDict

from sqlalchemy.dialects.mssql import BIT, CHAR, DATE, DECIMAL, INTEGER, SMALLINT, VARCHAR

type_metadata = OrderedDict([
    ('legal_status_code', [CHAR(1), 'string']),
    ('crime_degree', [CHAR(2), 'string']),
    ('gender', [CHAR(1), 'string']),
    ('agent_name', [VARCHAR(40), 'string']),
    ('agent_id', [INTEGER(), 'Int32']),
    ('date_of_birth', [DATE(), 'datetime64[ns]']),
    ('supervisor_id', [VARCHAR(8), 'string']),
    ('supervisor_name', [VARCHAR(35), 'string']),
    ('offender', [VARCHAR(60), 'string']),
    ('active_warrant', [BIT(), 'Int8']),
    ('legal_status', [VARCHAR(10), 'string']),
    ('last_successful_field_contact', [INTEGER(), 'Int32']),
    ('supervision_start_date', [DATE(), 'datetime64[ns]']),
    ('offense_description', [VARCHAR(80), 'string']),
    ('agency', [VARCHAR(30), 'string']),
    ('race', [VARCHAR(30), 'string']),
    ('special_supervision', [VARCHAR(30), 'string']),
    ('last_attempted_field_contact', [INTEGER(), 'Int32']),
    ('in_compliance', [BIT(), 'Int8']),
    ('last_office_contact', [INTEGER(), 'Int32']),
    ('offense_code', [CHAR(1), 'string']),
    ('primary_offense', [VARCHAR(30), 'string']),
    ('region', [SMALLINT(), 'Int8']),
    ('offender_location', [VARCHAR(40), 'string']),
    ('offender_id', [INTEGER(), 'Int32']),
    ('legal_status_description', [VARCHAR(30), 'string']),
    ('earned_compliance_credit', [DATE(), 'datetime64[ns]']),
    ('city', [VARCHAR(30), 'string']),
    ('county', [VARCHAR(12), 'string']),
    ('score', [DECIMAL(5, 2), 'Int16']),
    ('address_start_date', [DATE(), 'string']),
    ('zip', [INTEGER(), 'Int']),
    ('state', [CHAR(2), 'string']),
    ('address', [VARCHAR(50), 'string']),
    ('cautions', [VARCHAR(3500), 'string']),
    ('address_type', [VARCHAR(20), 'string']),
    ('x', [DECIMAL(24, 20), 'Int32']),
    ('y', [DECIMAL(24, 20), 'Int32']),
    ('address_type_code', [CHAR(1), 'string']),
    ('last_field_contact', [DATE(), 'datetime64[ns]']),
    ('field_contact_result', [VARCHAR(20), 'string']),
    ('alerts', [VARCHAR(500), 'string']),
    ('unit', [VARCHAR(30), 'string']),
    ('standard_of_supervision', [VARCHAR(5), 'string']),
    ('employer', [VARCHAR(30), 'string']),
    ('employer_phone', [VARCHAR(12), 'string']),
    ('employer_address', [VARCHAR(60), 'string']),
    ('offender_phone', [VARCHAR(18), 'string']),
    ('gang_name', [VARCHAR(30), 'string']),
    ('gang_type', [VARCHAR(20), 'string']),
    ('gang_id', [INTEGER(), 'Int32']),
    ('gang_type_id', [INTEGER(), 'Int32']),
])

DATA_TYPES = {header: value[1] for (header, value) in type_metadata.items()}
SQL_TYPES = {header: value[0] for (header, value) in type_metadata.items()}

SPECIAL_SUPERVISION = [
    'CCC',
    'COMP',
    'DEP',
    'DORA',
    'DRUG CT',
    'ECR',
    'EM',
    'FOSI',
    'FUG',
    'GPS',
    'IG INT',
    'INCAR',
    'MIO',
    'PVP',
    'RESID',
    'SO',
    'SO-A',
    'SO-B',
    'SO-C',
]

AGENT_SQL_TYPES = {
    'id': INTEGER(),
    'value': VARCHAR(128),
    'agent_id': VARCHAR(64),
    'supervisor_id': VARCHAR(64),
    'supervisor_name': VARCHAR(128),
}

File: src/Models/Schema.cs
using System;

namespace parole.Models
{
  public class Schema
  {
    public int offender_id { get; set; }
    public char gender { get; set; }
    public int region { get; set; }
    public string agency { get; set; }
    public string supervisor_id { get; set; }
    public string supervisor_name { get; set; }
    public string agent_name { get; set; }
    public string offender { get; set; }
    public DateTime date_of_birth { get; set; }
    public string race { get; set; }
    public string legal_status { get; set; }
    public char legal_status_code { get; set; }
    public string legal_status_description { get; set; }
    public DateTime supervision_start_date { get; set; }
    public string offender_location { get; set; }
    public DateTime address_start_date { get; set; }
    public string address { get; set; }
    public string unit { get; set; }
    public string city { get; set; }
    public string state { get; set; }
    public int zip { get; set; }
    public char address_type_code { get; set; }
    public string address_type { get; set; }
    public decimal x { get; set; }
    public decimal y { get; set; }
    public double score { get; set; }
    public string offender_phone { get; set; }
    public string standard_of_supervision { get; set; }
    public int last_office_contact { get; set; }
    public int last_successful_field_contact { get; set; }
    public DateTime last_field_contact { get; set; }
    public string field_contact_result { get; set; }
    public char offense_code { get; set; }
    public string primary_offense { get; set; }
    public string crime_degree { get; set; }
    public string offense_description { get; set; }
    public DateTime earned_compliance_credit { get; set; }
    public bool active_warrant { get; set; }
    public int gang_id { get; set; }
    public string gang_name { get; set; }
    public int gang_type_id { get; set; }
    public string gang_type { get; set; }
    public int agent_id { get; set; }
    public int last_attempted_field_contact { get; set; }
    public bool in_compliance { get; set; }
    public string alerts { get; set; }
    public string cautions { get; set; }
    public string employer { get; set; }
    public string employer_address { get; set; }
    public string employer_phone { get; set; }
    public string county { get; set; }
    public bool ccc { get; set; }
    public bool comp { get; set; }
    public bool dep { get; set; }
    public bool dora { get; set; }
    public bool drugct { get; set; }
    public bool ecr { get; set; }
    public bool em { get; set; }
    public bool fosi { get; set; }
    public bool fug { get; set; }
    public bool gps { get; set; }
    public bool igint { get; set; }
    public bool incar { get; set; }
    public bool mio { get; set; }
    public bool pvp { get; set; }
    public bool resid { get; set; }
    public bool so { get; set; }
    public bool soa { get; set; }
    public bool sob { get; set; }
    public bool soc { get; set; }
  }
}
